src/main/kotlin/com/hedvig/accountservice/accounts/model/Entry.kt

```diff
@@ -13,68 +14,91 @@ import java.util.*
-            amount = entryDto.amount,
-            type = entryDto.type,
-            source = entryDto.source,
-            reference = entryDto.reference,
-            matchesEntryId = entryDto.matchesEntryId,
-            title = entryDto.title,
-            comment = entryDto.comment,
-            addedBy = entryDto.addedBy ?: "system",
-            addedAt = addedAt
-        )
+  companion object {
+    fun from(entryDto: EntryRequestDto, memberId: String, addedAt: Instant = Instant.now()): Entry = Entry(
+      memberId = memberId,
+      fromDate = entryDto.fromDate,
+      amount = entryDto.amount,
+      type = entryDto.type,
+      source = entryDto.source,
+      reference = entryDto.reference,
+      matchesEntryId = entryDto.matchesEntryId,
+      title = entryDto.title,
+      comment = entryDto.comment,
+      addedBy = entryDto.addedBy ?: "system",
+      addedAt = addedAt
+    )
 
-        fun from(event: ScheduleSubscriptionEvent, period: YearMonth, productSubscription: InsuranceProductSubscriptionDto, timestamp: Instant): Entry = Entry(
-            memberId = event.memberId,
-            fromDate = period.atDay(1),
-            amount = productSubscription.subscription,
-            type = EntryType.SUBSCRIPTION,
-            source = "product-pricing",
-            reference = productSubscription.productId.toString(),
-            comment = productSubscription.productType.name,
-            addedAt = timestamp,
-            addedBy = event.scheduledBy
-        )
+    fun from(
+      event: ScheduleSubscriptionEvent,
+      period: YearMonth,
+      productSubscription: InsuranceProductSubscriptionDto,
+      timestamp: Instant
+    ): Entry = Entry(
+      memberId = event.memberId,
+      fromDate = period.atDay(1),
+      amount = productSubscription.subscription,
+      type = EntryType.SUBSCRIPTION,
+      source = "product-pricing",
+      reference = productSubscription.productId.toString(),
+      comment = productSubscription.productType.name,
+      addedAt = timestamp,
+      addedBy = event.scheduledBy
+    )
 
-        fun from(event: ChargeRequestedEvent, fromDate: LocalDate, transactionId: UUID, timestamp: Instant): Entry = Entry(
-            memberId = event.memberId,
-            fromDate = fromDate,
-            amount = event.amount.negate(),
-            type = EntryType.CHARGE,
-            source = "payment-service",
-            reference = transactionId.toString(),
-            addedBy = event.requestedBy,
-            addedAt = timestamp
-        )
+    fun from(
+      event: ScheduleSubscriptionEvent,
+      period: YearMonth,
+      discount: CampaignDiscount,
+      timestamp: Instant
+    ): Entry = Entry(
+      memberId = event.memberId,
+      fromDate = period.atDay(1),
+      amount = discount.amount.negate(),
+      type = EntryType.fromCampaignDiscount(discount),
+      source = "campaign-service",
+      reference = discount.reference ?: discount.memberId,
```

Comment:
It would be nice if this logic was more explicit (inspired by the EntryType)
```suggestion
      reference = when (discount.type) {
         DiscountType.FREE_MONTH -> discount.reference,
         DiscountType.REFERRAL -> discount.memeberId
      }
```